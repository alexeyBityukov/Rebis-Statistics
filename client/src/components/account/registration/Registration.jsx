import React from 'react';
import { withStyles, createStyles } from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import ErrorMessage from '../../messages/ErrorMessage';
import TextField from '../../textFields';


const listFields = {
  email: 'email',
  password: 'password',
  repeatPassword: 'repeatPassword',
};

const styles = ({ spacing }) => createStyles({
  paper: {
    margin: spacing.unit,
    padding: spacing.unit,
  },
});

const validate = fields => (values) => {
  const errors = {};
  Object.values(fields).forEach((key) => {
    if (!values[key]) errors[key] = 'requiredField';
  });
  if (values.email && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) errors.email = 'invalidEmail';

  if (values.password !== values.repeatPassword) {
    errors.password = 'differentPasswords';
    errors.repeatPassword = 'differentPasswords';
  }

  return errors;
};

export default connect(
  state => ({ lang: state.main.lang }),
)(withStyles(styles)(reduxForm(
  {
    form: 'registration',
    validate: validate(listFields),
  },
)((props) => {
  const {
    classes,
    lang: { registration },
    handleSubmit,
    errorMessage,
  } = props;
  return (
    <Paper className={classes.paper}>
      <form autoComplete="off" onSubmit={handleSubmit}>
        <Field
          name={listFields.email}
          required
          label={registration.emailFieldLabel}
          component={TextField}
        />
        <Field
          name={listFields.password}
          required
          label={registration.passwordFieldLabel}
          component={TextField}
        />
        <Field
          name={listFields.repeatPassword}
          required
          label={registration.repeatPasswordFieldLabel}
          component={TextField}
        />
        <ErrorMessage label={errorMessage} />
        <Button
          color="primary"
          type="submit"
        >
          {registration.registerButtonLabel}
        </Button>
      </form>
    </Paper>
  );
})));
