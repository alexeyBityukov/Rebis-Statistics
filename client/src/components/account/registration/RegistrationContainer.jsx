import React, { Component } from 'react';
import Registration from './Registration';

export default class extends Component {
  state = {
    fields: {
    }
  };

  handleOnChange = (fieldName) => {
    return (event) => {
      event.preventDefault();
      this.setState({
        fields: {
          ...this.state.fields,
          [fieldName]: event.target.value,
        }
      });
    }
  };

  handleOnClick = (event) => {
    event.preventDefault();
    console.log('send req on save');
    console.log(this.state);
  };

  render() {
    return <Registration
      handleOnChange={this.handleOnChange}
      handleOnClick={this.handleOnClick}
      fields={this.state.fields}
    />;
  }
};
