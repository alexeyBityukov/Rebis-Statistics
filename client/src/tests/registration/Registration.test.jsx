import React from 'react';
import { configure, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import { Registration, validate } from '../../components/account/registration/Registration';

configure({ adapter: new Adapter() });

const mockStore = configureMockStore();

let component;
let store;
const registrationFormLabel = {
  emailFieldLabel: 'email',
  passwordFieldLabel: 'pass',
  repeatPasswordFieldLabel: 'repeat pass',
};

beforeEach(() => {
  store = mockStore({});
  component = shallow((<Registration
    store={store}
    lang={registrationFormLabel}
    handleSubmit={() => true}
    errorMessage=""
  />));
});

describe('<Registration>', () => {
  it('should shallow render correctly', () => {
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should shallow render with error message  correctly', () => {
    const message = 'message';

    component = shallow((<Registration
      store={store}
      lang={registrationFormLabel}
      handleSubmit={() => true}
      errorMessage={message}
    />));

    expect(shallowToJson(component)).toMatchSnapshot();
    expect(component.props().errorMessage).toEqual(message);
  });

  it('should validate correct', () => {
    const correctFieldsName = { field: 'field', email: 'email' };
    const correctFieldsValue = { field: 'value', email: 'test@ya.ru' };
    const emptyFieldsValue = { field: '', email: '' };
    const errorRequiredFields = { field: 'requiredField', email: 'requiredField' };
    const invalidEmailValue = { field: 'value', email: 'test' };
    const errorInvalidEmail = { email: 'invalidEmail' };

    expect(validate(correctFieldsName)(correctFieldsValue)).toEqual({});
    expect(validate(correctFieldsName)(emptyFieldsValue)).toEqual(errorRequiredFields);
    expect(validate(correctFieldsName)(invalidEmailValue)).toEqual(errorInvalidEmail);
  });
});
