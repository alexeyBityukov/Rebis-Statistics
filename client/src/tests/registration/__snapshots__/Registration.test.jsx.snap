// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration> should shallow render correctly 1`] = `
<ReduxForm
  classes={
    Object {
      "paper": "ReduxForm-paper-1",
    }
  }
  destroyOnUnmount={true}
  enableReinitialize={false}
  errorMessage=""
  forceUnregisterOnUnmount={false}
  form="registration"
  getFormState={[Function]}
  handleSubmit={[Function]}
  keepDirtyOnReinitialize={false}
  lang={
    Object {
      "emailFieldLabel": "email",
      "passwordFieldLabel": "pass",
      "repeatPasswordFieldLabel": "repeat pass",
    }
  }
  persistentSubmitErrors={false}
  pure={true}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
  validate={[Function]}
/>
`;

exports[`<Registration> should shallow render with error message  correctly 1`] = `
<ReduxForm
  classes={
    Object {
      "paper": "ReduxForm-paper-1",
    }
  }
  destroyOnUnmount={true}
  enableReinitialize={false}
  errorMessage="message"
  forceUnregisterOnUnmount={false}
  form="registration"
  getFormState={[Function]}
  handleSubmit={[Function]}
  keepDirtyOnReinitialize={false}
  lang={
    Object {
      "emailFieldLabel": "email",
      "passwordFieldLabel": "pass",
      "repeatPasswordFieldLabel": "repeat pass",
    }
  }
  persistentSubmitErrors={false}
  pure={true}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
  validate={[Function]}
/>
`;

exports[`<Registration> should shallow render with error message  correctly 2`] = `
<ReduxForm
  classes={
    Object {
      "paper": "ReduxForm-paper-1",
    }
  }
  destroyOnUnmount={true}
  enableReinitialize={false}
  errorMessage="message"
  forceUnregisterOnUnmount={false}
  form="registration"
  getFormState={[Function]}
  handleSubmit={[Function]}
  keepDirtyOnReinitialize={false}
  lang={
    Object {
      "emailFieldLabel": "email",
      "passwordFieldLabel": "pass",
      "repeatPasswordFieldLabel": "repeat pass",
    }
  }
  persistentSubmitErrors={false}
  pure={true}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
  validate={[Function]}
/>
`;
